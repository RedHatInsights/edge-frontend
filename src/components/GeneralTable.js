import React, { Fragment, useEffect, useState } from 'react';
import {
  Table,
  TableHeader,
  TableBody,
  sortable,
} from '@patternfly/react-table';
import {
  EmptyState,
  EmptyStateIcon,
  EmptyStateSecondaryActions,
  Title,
  Button,
  Spinner,
  Bullseye,
} from '@patternfly/react-core';
import { PrimaryToolbar } from '@redhat-cloud-services/frontend-components/PrimaryToolbar';
import {
  isEmptyFilters,
  constructActiveFilters,
  onDeleteFilter,
} from '../constants';
import { useDispatch } from 'react-redux';
import { PlusCircleIcon, SearchIcon } from '@patternfly/react-icons';
import {
  transformFilters,
  transformPaginationParams,
  transformSort,
} from '../Routes/ImageManager/constants';
import PropTypes from 'prop-types';

const GeneralTable = ({
  tableData,
  columnNames,
  createRows,
  emptyStateMessage,
  emptyStateActionMessage,
  emptyStateAction,
  defaultSort,
  loadTableData,
  filters,
  filterDep,
  clearFilters,
  actionResolver,
  areActionsDisabled,
  filterConfig,
  activeFilters,
  dispatchActiveFilters,
  defaultFilters,
  perPage
}) => {
  const [sortBy, setSortBy] = useState(defaultSort);
  const [pagination, setPagination] = useState({ page: 1, perPage });
  const dispatch = useDispatch();

  const columns = columnNames.map((columnName) => ({
    title: columnName.title,
    type: columnName.type,
    transforms: toShowSort ? [] : columnName.sort ? [sortable] : [],
  }));

  const { count, data, isLoading, hasError } = tableData;

  const toShowSort = isLoading || hasError || (!count?.length && hasFilters);
  useEffect(() => {
    loadTableData(dispatch, {
      ...transformFilters(filters),
      ...transformPaginationParams(pagination),
      ...transformSort({
        direction: sortBy.direction,
        name: columns[sortBy.index].type,
      }),
    });
  }, [
    pagination.perPage,
    pagination.page,
    sortBy.index,
    sortBy.direction,
    ...filterDep,
  ]);

  const hasFilters = Object.keys(filters).some((filterKey) => filterKey);

  let rows = [
    {
      heightAuto: true,
      cells: [
        {
          props: { colSpan: 8 },
          title: (
            <Bullseye>
              <EmptyState variant="small">
                <EmptyStateIcon icon={Spinner} />
              </EmptyState>
            </Bullseye>
          ),
        },
      ],
    },
  ];

  if (isLoading === false && hasError === false) {
    if (!count?.length && !hasFilters) {
      rows = [
        {
          heightAuto: true,
          cells: [
            {
              props: { colSpan: 8 },
              title: (
                <Bullseye>
                  <EmptyState variant="small">
                    <EmptyStateIcon icon={PlusCircleIcon} />
                    <Title headingLevel="h2" size="lg">
                      {emptyStateMessage}
                    </Title>
                    {emptyStateActionMessage ? (
                      <Button
                        onClick={emptyStateAction}
                        isDisabled={isLoading !== false}
                      >
                        {emptyStateActionMessage}
                      </Button>
                    ) : null}
                  </EmptyState>
                </Bullseye>
              ),
            },
          ],
        },
      ];
    }
    if (!count?.length && hasFilters) {
      rows = [
        {
          heightAuto: true,
          cells: [
            {
              props: { colSpan: 8 },
              title: (
                <Bullseye>
                  <EmptyState variant="small">
                    <EmptyStateIcon icon={SearchIcon} />
                    <Title headingLevel="h2" size="lg">
                      No match found
                    </Title>
                    <EmptyStateSecondaryActions>
                      <Button onClick={clearFilters} variant="link">
                        Clear all filters
                      </Button>
                    </EmptyStateSecondaryActions>
                  </EmptyState>
                </Bullseye>
              ),
            },
          ],
        },
      ];
    }

    if (data?.length) {
      rows = createRows(data);
    }
  }

  const handleSort = (_event, index, direction) => {
    setSortBy({ index, direction });
  };

  return (
    <Fragment>
      <PrimaryToolbar
        filterConfig={filterConfig}
        pagination={{
          itemCount: count,
          ...pagination,
          onSetPage: (_evt, newPage) =>
            setPagination({ ...pagination, page: newPage }),
          onPerPageSelect: (_evt, newPerPage) =>
            setPagination({ page: 1, perPage: newPerPage }),
        }}
        activeFiltersConfig={{
          filters: isEmptyFilters(activeFilters)
            ? constructActiveFilters(activeFilters)
            : [],
          onDelete: (_event, itemsToRemove, isAll) => {
            if (isAll) {
              dispatchActiveFilters({
                type: 'DELETE_FILTER',
                payload: defaultFilters,
              });
            } else {
              dispatchActiveFilters({
                type: 'DELETE_FILTER',
                payload: onDeleteFilter(activeFilters, itemsToRemove),
              });
            }
          },
        }}
        dedicatedAction={
          <Button onClick={emptyStateAction} isDisabled={isLoading !== false}>
            {emptyStateActionMessage}
          </Button>
        }
      />
      <Table
        variant="compact"
        aria-label="Manage Images table"
        sortBy={sortBy}
        onSort={handleSort}
        actionResolver={actionResolver}
        areActionsDisabled={areActionsDisabled}
        cells={columns}
        rows={rows}
      >
        <TableHeader />
        <TableBody />
      </Table>
    </Fragment>
  );
};

GeneralTable.propTypes = {
  tableData: PropTypes.object.isRequired,
  columnNames: PropTypes.array.isRequired,
  createRows: PropTypes.func.isRequired,
  emptyStateMessage: PropTypes.string.isRequired,
  emptyStateActionMessage: PropTypes.string,
  emptyStateAction: PropTypes.func,
  defaultSort: PropTypes.object.isRequired,
  clearFilters: PropTypes.func.isRequired,
  filters: PropTypes.array.isRequired,
  loadTableData: PropTypes.func.isRequired,
  filterDep: PropTypes.array.isRequired,
  actionResolver: PropTypes.func.isRequired,
  areActionsDisabled: PropTypes.func.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }).isRequired,
  setPagination: PropTypes.func.isRequired,
  filterConfig: PropTypes.shape({ items: PropTypes.array }),
  defaultFilters: PropTypes.shape({
    name: PropTypes.object,
    distribution: PropTypes.object,
    status: PropTypes.object,
  }),
  activeFilters: PropTypes.shape({
    name: PropTypes.object,
    distribution: PropTypes.object,
    status: PropTypes.object,
  }),
  dispatchActiveFilters: PropTypes.func.isRequired,
};

export default GeneralTable;
