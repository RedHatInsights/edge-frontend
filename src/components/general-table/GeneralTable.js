import React, { useState, useEffect } from 'react';
import ToolbarHeader from './ToolbarHeader';
import ToolbarFooter from './ToolbarFooter';
import createFilterValues from '../../components/general-table/createFilterValues';
import {
  Table,
  TableHeader,
  TableBody,
  sortable,
} from '@patternfly/react-table';
import { Skeleton } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import CustomEmptyState from '../Empty';
import { useDispatch } from 'react-redux';
import { transformSort } from '../../utils';
import BulkSelect from './BulkSelect';
import { useHistory, useLocation, useNavigate } from 'react-router-dom';
import { stateToUrlSearch } from '../../utils';

const filterParams = (chipsArray) => {
  const filterParamsObj =
    chipsArray.length > 0
      ? chipsArray.reduce((acc, filter) => {
          if (acc[filter.key.toLowerCase()]) {
            const returnAcc =
              typeof acc[filter.key.toLowerCase()] === 'string'
                ? [acc[filter.key.toLowerCase()]]
                : [...acc[filter.key.toLowerCase()]];
            return {
              ...acc,
              [filter.key.toLowerCase()]: [
                ...returnAcc,
                filter.apiName ? filter.apiName : filter.value,
              ],
            };
          } else {
            return {
              ...acc,
              [filter.key.toLowerCase()]: filter.apiName
                ? filter.apiName
                : filter.value || filter.label,
            };
          }
        }, {})
      : {};
  return filterParamsObj;
};

const GeneralTable = ({
  historyProp,
  locationProp,
  navigateProp,
  apiFilterSort,
  urlParam,
  filters,
  loadTableData,
  tableData,
  columnNames,
  rows,
  toolbarButtons,
  actionResolver,
  areActionsDisabled,
  defaultSort,
  emptyFilterState,
  toggleButton,
  toggleAction,
  toggleState,
  hasCheckbox = false,
  skeletonRowQuantity,
  selectedItems,
  initSelectedItems,
  kebabItems,
  hasModalSubmitted,
  setHasModalSubmitted,
  isUseApi,
  hasRadio = false,
  setRadioSelection,
  isFooterFixed = false,
}) => {
  const defaultCheckedRows = initSelectedItems ? initSelectedItems : [];
  const [filterValues, setFilterValues] = useState(createFilterValues(filters));
  const [chipsArray, setChipsArray] = useState([]);
  const [sortBy, setSortBy] = useState(defaultSort);
  const [perPage, setPerPage] = useState(50);
  const [page, setPage] = useState(1);
  const [checkedRows, setCheckedRows] = useState(defaultCheckedRows);
  const dispatch = useDispatch();
  const history = historyProp
    ? historyProp()
    : useHistory
    ? useHistory()
    : null;
  const navigate = navigateProp
    ? navigateProp()
    : useNavigate
    ? useNavigate()
    : null;
  const { pathname, search } = locationProp ? locationProp() : useLocation();

  useEffect(() => {
    // Add or remove has_filters param depending on whether filters are present
    if (
      !search.includes('create_image=true') &&
      !search.includes('update_image=true')
    ) {
      const a = {
        pathname,
        search: stateToUrlSearch(
          'has_filters=true',
          chipsArray.length > 0,
          search
        ),
      };
      if (navigateProp) {
        navigate({ ...a, replace: true });
      } else {
        history.replace(a);
      }
    }

    const query = apiFilterSort
      ? {
          ...filterParams(chipsArray),
          limit: perPage,
          offset: (page - 1) * perPage,
          ...transformSort({
            direction: sortBy.direction,
            name: columns[sortBy.index].type,
          }),
        }
      : null;

    if (query?.status === 'updateAvailable') {
      delete query.status;
      query.update_available = 'true';
    }

    if (query && query['image status'] === 'updateAvailable') {
      delete query['image status'];
      query.update_available = 'true';
    }

    if (isUseApi) {
      loadTableData(query);
      return;
    }

    apiFilterSort && urlParam
      ? loadTableData(dispatch, urlParam, query)
      : apiFilterSort
      ? loadTableData(dispatch, query)
      : null;
  }, [chipsArray, perPage, page, sortBy, hasModalSubmitted]);

  useEffect(() => {
    setCheckedRows(defaultCheckedRows);
  }, [hasModalSubmitted]);

  useEffect(() => {
    selectedItems && selectedItems(checkedRows);
    hasModalSubmitted && setHasModalSubmitted(false);
  }, [checkedRows]);

  const { count, isLoading, hasError } = tableData;

  //Used for repos until the api can sort and filter
  const filteredByName = () => {
    const activeFilters = filterValues.filter(
      (filter) =>
        (filter?.type === 'text' && filter?.value !== '') ||
        (filter?.type === 'checkbox' &&
          filter?.value.find((checked) => checked.isChecked))
    );
    const filteredArray = rows.filter((row) => {
      if (activeFilters.length > 0) {
        return activeFilters?.every((filter) => {
          if (filter.type === 'text') {
            return row.noApiSortFilter[
              columnNames.findIndex((row) => row.title === filter.label)
            ]
              .toLowerCase()
              .includes(filter.value.toLowerCase());
          } else if (filter.type === 'checkbox') {
            return filter.value.some(
              (value) =>
                value.isChecked &&
                row.noApiSortFilter[
                  columnNames.findIndex((row) => row.title === filter.label) - 1
                ].toLowerCase() === value.value.toLowerCase()
            );
          }
        });
      } else {
        return row;
      }
    });
    return filteredArray;
  };

  const filteredByNameRows = !apiFilterSort && filteredByName();

  //non-api sort function
  const sortedByDirection = (rows) =>
    rows.sort((a, b) => {
      const index = hasCheckbox ? sortBy.index - 1 : sortBy.index;
      return typeof a?.noApiSortFilter[index] === 'number'
        ? sortBy.direction === 'asc'
          ? a?.noApiSortFilter[index] - b?.noApiSortFilter[index]
          : b?.noApiSortFilter[index] - a?.noApiSortFilter[index]
        : sortBy.direction === 'asc'
        ? a?.noApiSortFilter[index].localeCompare(
            b?.noApiSortFilter[index],
            undefined,
            { sensitivity: 'base' }
          )
        : b?.noApiSortFilter[index].localeCompare(
            a?.noApiSortFilter[index],
            undefined,
            { sensitivity: 'base' }
          );
    });

  const nonApiCount = !apiFilterSort
    ? sortedByDirection(filteredByNameRows)?.length
    : 0;

  const handleSort = (_event, index, direction) => {
    index = hasCheckbox ? index - 1 : index;
    setSortBy({ index, direction });
  };

  const toShowSort =
    isLoading || hasError || (count?.length > 0 && filters.length > 0);

  const columns = columnNames.map((columnName) => ({
    title: columnName.title,
    type: columnName.type,
    transforms: toShowSort ? [] : columnName.sort ? [sortable] : [],
    columnTransforms: columnName.columnTransforms
      ? columnName.columnTransforms
      : [],
  }));

  const filteredRows = apiFilterSort
    ? rows
    : rows.length > 0
    ? sortedByDirection(filteredByNameRows).slice(
        (page - 1) * perPage,
        (page - 1) * perPage + perPage
      )
    : rows;

  const checkboxRows = () =>
    filteredRows.map((row) =>
      checkedRows.some((checkedRow) => checkedRow.id === row.rowInfo.id)
        ? {
            ...row,
            selected: true,
          }
        : {
            ...row,
            selected: false,
          }
    );

  const handleSelect = (_event, isSelecting, rowIndex) => {
    setCheckedRows((prevState) => {
      return isSelecting
        ? [...prevState, { ...filteredRows[rowIndex].rowInfo }]
        : prevState.filter(
            (row) => row.id !== filteredRows[rowIndex].rowInfo.id
          );
    });
  };

  const handlePageSelect = () => {
    setCheckedRows((prevState) => {
      const checkedIds = prevState.map((row) => row.id);
      const rowIsNotIncluded = (id) => !checkedIds.includes(id);

      const newRows = [];
      filteredRows.forEach((filtered) => {
        if (rowIsNotIncluded(filtered.rowInfo.id)) {
          newRows.push({
            ...filtered.rowInfo,
          });
        }
      });

      return [...prevState, ...newRows];
    });
  };

  const handleBulkSelect = () => {
    setCheckedRows(
      rows.map((row) => ({
        ...row.rowInfo,
      }))
    );
  };

  const handleNoneSelect = () => {
    setCheckedRows([]);
  };

  const loadingRows = (perPage) =>
    [...Array(skeletonRowQuantity ?? perPage)].map(() => ({
      cells: columnNames.map(() => ({ title: <Skeleton width="100%" /> })),
    }));

  const emptyFilterView = () => {
    hasCheckbox = false;
    return [
      {
        heightAuto: true,
        cells: [
          {
            props: {
              colSpan: 8,
            },
            title: (
              <CustomEmptyState
                data-testid="general-table-empty-state-no-match"
                bgColor="white"
                icon={emptyFilterState?.icon ?? 'search'}
                title={emptyFilterState?.title ?? 'No match found'}
                body={emptyFilterState?.body ?? ''}
                secondaryActions={
                  toggleAction
                    ? []
                    : [
                        {
                          title: 'Clear all filters',
                          onClick: () =>
                            setFilterValues(createFilterValues(filters)),
                        },
                      ]
                }
              />
            ),
          },
        ],
      },
    ];
  };

  const tableRows = isLoading
    ? loadingRows(perPage)
    : !filteredRows?.length > 0
    ? emptyFilterView()
    : hasCheckbox
    ? checkboxRows()
    : filteredRows;

  const onSelect = () => {
    if (isLoading) {
      return null;
    }

    if (hasCheckbox) {
      // change to handleCheckboxSelect
      return handleSelect;
    }

    if (hasRadio) {
      return (_event, _isSelecting, rowIndex) => {
        const versionData = tableRows[rowIndex];
        setRadioSelection(versionData);
      };
    }
  };

  return (
    <>
      <ToolbarHeader
        count={apiFilterSort ? count : nonApiCount}
        toolbarButtons={toolbarButtons}
        filters={filters}
        filterValues={filterValues}
        setFilterValues={setFilterValues}
        chipsArray={chipsArray}
        setChipsArray={setChipsArray}
        isLoading={isLoading}
        perPage={perPage}
        setPerPage={setPerPage}
        page={page}
        setPage={setPage}
        toggleButton={toggleButton}
        toggleAction={toggleAction}
        toggleState={toggleState}
        kebabItems={kebabItems}
      >
        {!isLoading && hasCheckbox ? (
          <BulkSelect
            checkedRows={checkedRows}
            handleBulkSelect={handleBulkSelect}
            handlePageSelect={handlePageSelect}
            handleNoneSelect={handleNoneSelect}
            displayedRowsLength={filteredRows.length}
            perPage={perPage}
            total={apiFilterSort ? count : nonApiCount}
          />
        ) : null}
      </ToolbarHeader>
      <Table
        data-testid="general-table-testid"
        variant="compact"
        aria-label="General Table Component"
        sortBy={hasCheckbox ? { ...sortBy, index: sortBy.index + 1 } : sortBy}
        onSort={handleSort}
        actionResolver={actionResolver ? actionResolver : null}
        areActionsDisabled={areActionsDisabled}
        cells={columns}
        rows={tableRows}
        onSelect={onSelect()}
        canSelectAll={false}
        selectVariant={hasRadio ? 'radio' : hasCheckbox ? 'checkbox' : ''}
      >
        <TableHeader />
        <TableBody />
      </Table>
      <ToolbarFooter
        isLoading={isLoading}
        count={apiFilterSort ? count : nonApiCount}
        setFilterValues={setFilterValues}
        perPage={perPage}
        setPerPage={setPerPage}
        page={page}
        setPage={setPage}
        isFooterFixed={isFooterFixed}
      />
    </>
  );
};

GeneralTable.propTypes = {
  historyProp: PropTypes.func,
  locationProp: PropTypes.func,
  navigateProp: PropTypes.func,
  apiFilterSort: PropTypes.bool,
  filters: PropTypes.array,
  urlParam: PropTypes.string,
  loadTableData: PropTypes.func,
  tableData: PropTypes.object,
  columnNames: PropTypes.array,
  rows: PropTypes.array,
  actionResolver: PropTypes.func,
  areActionsDisabled: PropTypes.func,
  defaultSort: PropTypes.object,
  toolbarButtons: PropTypes.array,
  toggleButton: PropTypes.array,
  toggleAction: PropTypes.func,
  toggleState: PropTypes.number,
  hasCheckbox: PropTypes.bool,
  skeletonRowQuantity: PropTypes.number,
  emptyFilterState: PropTypes.object,
  selectedItems: PropTypes.func,
  kebabItems: PropTypes.array,
  hasModalSubmitted: PropTypes.bool,
  setHasModalSubmitted: PropTypes.func,
  initSelectedItems: PropTypes.array,
  isUseApi: PropTypes.bool,
  hasToolbar: PropTypes.bool,
  hasRadio: PropTypes.bool,
  setRadioSelection: PropTypes.func,
  isFooterFixed: PropTypes.bool,
};

GeneralTable.defaultProps = {
  hasModalSubmitted: false,
  setHasModalSubmitted: () => {},
};

export default GeneralTable;
