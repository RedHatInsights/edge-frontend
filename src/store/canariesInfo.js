import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { LOAD_CANARIES_INFO } from './action-types';

const initialState = {};

const loadCanariesInfoPending = (state) => ({
  ...state,
  canariesInfo: [],
  isLoading: true,
});
const loadCanariesInfoFulfilled = (state, { payload }) => ({
  ...state,
  canariesInfo: payload?.results || [],
  isLoading: false,
});

export default applyReducerHash(
  {
    [`${LOAD_CANARIES_INFO}_PENDING`]: loadCanariesInfoPending,
    [`${LOAD_CANARIES_INFO}_FULFILLED`]: loadCanariesInfoFulfilled,
  },
  initialState
);
