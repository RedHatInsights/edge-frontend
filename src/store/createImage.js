import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { CREATE_NEW_IMAGE } from './action-types';

const initialState = {};

const createNewImagePending = () => ({
  isLoading: true,
  hasError: false,
  error: null,
});

const createNewImageFulfilled = () => {
  return {
    isLoading: false,
    hasError: false,
    error: null,
  };
};

const createNewImageRejected = (_state, { payload }) => ({
  isLoading: false,
  hasError: true,
  error: payload,
});

const createNewImageReset = () => ({
  isLoading: false,
  hasError: false,
  error: null,
});

export default applyReducerHash(
  {
    [`${CREATE_NEW_IMAGE}_PENDING`]: createNewImagePending,
    [`${CREATE_NEW_IMAGE}_FULFILLED`]: createNewImageFulfilled,
    [`${CREATE_NEW_IMAGE}_REJECTED`]: createNewImageRejected,
    [`${CREATE_NEW_IMAGE}_RESET`]: createNewImageReset,
  },
  initialState
);
