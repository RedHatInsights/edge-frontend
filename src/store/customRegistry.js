import ReducerRegistry from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

export const createRegistry = (
  initialState = {},
  middleware = [],
  composeEnhancersDefault
) => {
  return new ReducerRegistry(
    initialState,
    [...middleware],
    composeEnhancersDefault
  );
};
