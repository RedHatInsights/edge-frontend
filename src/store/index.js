import { createContext } from 'react';
import { ReducerRegistry } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import { polling } from './polling';

export const RegistryContext = createContext({
  getRegistry: () => {},
});

export function init(...middleware) {
  return new ReducerRegistry({}, [
    promiseMiddleware,
    notificationsMiddleware({
      errorDescriptionKey: ['detail', 'stack'],
    }),
    polling(),
    ...middleware,
  ]);
}
