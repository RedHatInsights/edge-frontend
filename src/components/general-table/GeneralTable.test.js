import React from 'react';
import GeneralTable from './GeneralTable';

import { Provider } from 'react-redux';
import { init, RegistryContext } from '../../store';
import { render } from '@testing-library/react';
import logger from 'redux-logger';

import { MemoryRouter } from 'react-router-dom';
describe('General table', () => {
  it('should render correctly', async () => {
    const filters = [
      {
        label: 'Test 1',
        type: 'text',
      },
      {
        label: 'Test 2',
        type: 'checkbox',
        options: [
          { option: 'option 1', optionApiName: 'option-1' },
          { option: 'option 2' },
        ],
      },
    ];
    const columnNames = [
      { title: 'Name', type: 'name', sort: true },
      { title: 'Version', type: 'version', sort: false },
    ];
    const rows = [
      {
        cells: [{ Name: 'test-name' }, { Version: 'test-version' }],
      },
    ];
    const loadEdgeImages = jest.fn();
    const emptyStateAction = jest.fn();
    const actionResolver = jest.fn();
    const registry = init(logger);

    const { container, findByTestId, findByPlaceholderText } = render(
      <RegistryContext.Provider
        value={{
          getRegistry: () => registry,
        }}
      >
        <Provider store={registry.getStore()}>
          <GeneralTable
            apiFilterSort={false}
            filters={filters}
            loadTableData={loadEdgeImages}
            tableData={{
              count: 100,
              data: {},
              isLoading: false,
              hasError: false,
            }}
            columnNames={columnNames}
            rows={rows}
            emptyStateMessage="No data found"
            emptyStateActionMessage="Do something"
            emptyStateAction={emptyStateAction}
            actionResolver={actionResolver}
            areActionsDisabled={() => false}
            defaultSort={{ index: 5, direction: 'desc' }}
            toolbarButtons={[
              {
                title: 'test button',
                click: () => jest.fn(),
              },
            ]}
          />
        </Provider>
      </RegistryContext.Provider>,
      { wrapper: MemoryRouter }
    );

    const generalTableHeader = await findByTestId('toolbar-header-testid');
    const headerInput = await findByPlaceholderText('Filter by test 1');
    const generalTable = await findByTestId('general-table-testid');
    const generalTableFooter = await findByTestId('pagination-footer-test-id');

    expect(generalTableHeader).toBeDefined();
    expect(headerInput.value).toBeDefined();
    expect(generalTable).toBeDefined();
    expect(generalTableFooter).toBeDefined();
    expect(generalTableFooter.children[0].innerHTML).toEqual(
      '<b>1 - 1</b> of <b>1</b> '
    );

    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
