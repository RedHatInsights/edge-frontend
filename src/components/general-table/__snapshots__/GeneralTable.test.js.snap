// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General table should render correctly 1`] = `
<div
  class="pf-v5-c-toolbar"
  data-ouia-component-id="OUIA-Generated-Toolbar-1"
  data-ouia-component-type="PF5/Toolbar"
  data-ouia-safe="true"
  data-testid="toolbar-header-testid"
  id="toolbar-header"
  style="padding: 0px;"
>
  <div
    class="pf-v5-c-toolbar__content"
  >
    <div
      class="pf-v5-c-toolbar__content-section"
    >
      <div
        class="pf-v5-c-toolbar__item pf-u-mr-0"
        data-testid="filter-dropdown-testid"
      >
        <div
          class="pf-v5-c-select"
          data-ouia-component-id="OUIA-Generated-Select-single-1"
          data-ouia-component-type="PF5/Select"
          data-ouia-safe="true"
          style="width: 11rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Options menu"
            aria-labelledby=" pf-select-toggle-id-1"
            class="pf-v5-c-select__toggle"
            id="pf-select-toggle-id-1"
            type="button"
          >
            <div
              class="pf-v5-c-select__toggle-wrapper"
            >
              <span
                class="pf-v5-c-select__toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                  />
                </svg>
              </span>
              <span
                class="pf-v5-c-select__toggle-text"
              >
                Test 1
              </span>
            </div>
            <span
              class="pf-v5-c-select__toggle-arrow"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-toolbar__item"
        data-testid="filter-input-testid"
      >
        <div
          class="pf-v5-c-input-group"
        >
          <div
            class="pf-v5-c-input-group__item"
          >
            <div
              class="pf-v5-c-text-input-group"
              id="textInput1"
              type="search"
            >
              <div
                class="pf-v5-c-text-input-group__main pf-m-icon"
              >
                <span
                  class="pf-v5-c-text-input-group__text"
                >
                  <span
                    class="pf-v5-c-text-input-group__icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Select input for test 1"
                    class="pf-v5-c-text-input-group__text-input"
                    name="textInput1"
                    placeholder="Filter by test 1"
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-c-toolbar__item"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary"
          data-ouia-component-id="OUIA-Generated-Button-primary-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          test button
        </button>
      </div>
      <div
        class="pf-v5-c-toolbar__item pf-m-pagination pf-m-align-right"
      >
        <div
          class="pf-v5-c-pagination pf-m-compact"
          data-ouia-component-id="OUIA-Generated-Pagination-top-1"
          data-ouia-component-type="PF5/Pagination"
          data-ouia-safe="true"
          data-testid="pagination-header-test-id"
          id="pagination-options-menu-top-top-pagination"
          style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
        >
          <div
            class="pf-v5-c-pagination__total-items"
          >
            <b>
              1
               - 
              1
            </b>
             
            of
             
            <b>
              1
            </b>
             
          </div>
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
              id="pagination-options-menu-top-top-toggle"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__text"
              >
                <b>
                  1
                   - 
                  1
                </b>
                 
                of
                 
                <b>
                  1
                </b>
                 
              </span>
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
          <nav
            aria-label="Pagination"
            class="pf-v5-c-pagination__nav"
          >
            <div
              class="pf-v5-c-pagination__nav-control"
            >
              <button
                aria-disabled="true"
                aria-label="Go to previous page"
                class="pf-v5-c-button pf-m-plain pf-m-disabled"
                data-action="previous"
                data-ouia-component-id="OUIA-Generated-Button-plain-1"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                disabled=""
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="pf-v5-c-pagination__nav-control"
            >
              <button
                aria-disabled="true"
                aria-label="Go to next page"
                class="pf-v5-c-button pf-m-plain pf-m-disabled"
                data-action="next"
                data-ouia-component-id="OUIA-Generated-Button-plain-2"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                disabled=""
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </button>
            </div>
          </nav>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pf-v5-c-toolbar__content"
  >
    <div
      class="pf-v5-c-toolbar__content-section"
    >
      <div
        class="pf-v5-c-toolbar__item pf-m-chip-group pf-m-spacer-none"
      />
    </div>
  </div>
  <div
    class="pf-v5-c-toolbar__content pf-m-hidden"
    hidden=""
  >
    <div
      class="pf-v5-c-toolbar__group"
    />
  </div>
</div>
`;
